### PR TITLE
Add enumerable methods

### PR DESCRIPTION
**Pull request by appiersign**
In this pull request, I
- added Gemfile and rubocop installed
- added all required enumerable methods
- updated my_any? to with or without block
- added checks to handle range, hashes and list
- updated readme to show methods that were covered in this task

### CODE REVIEW

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @appiersign       
#### Great Job so far, but there are a few more things required. 
### General
- [ ] Lets make the README.md more descriptive by listing the methods you are working on.
- [ ] Lets activate stickler for this project.
- [ ] `my_each` 
-  returns an Enumerator if no block is given:
  `p [1, 2, 3].my_each` returns `#<Enumerator>`
- [ ] `my_each_with_index`  
- returns an Enumerator if no block is given:
  `p [1, 2, 3].my_each_with_index` returns `#<Enumerator>`
- it calls the given block once for each element in self
##### For example
      `my_each_output = ' ' `
      `my_each_output_two = ' ' `
      `block = proc { |y, x| my_each_output += "Hello: #{y}, in: #{x}\n" }`
      `block_two = proc { |y, x| my_each_output_two += "Hello: #{y}, in: #{x}\n" }`
      `[1,2,32,32,3,21,1,23,12].each_with_index(&block_two)`
      `[1,2,32,32,3,21,1,23,12].my_each_with_index(&block)`
       `p my_each_output ==  my_each_output_two ` this should return `true`
- [ ] `my_select`  
- returns an Enumerator if no block is given:
  `p [1, 2, 3].my_select` returns `#<Enumerator>`
- [ ] `my_all`  
- returns true if the block never returns false or nil
  `  block =  proc { |num| num < 200 } `
    `  p [2,32,32].my_all?(&block)  `should return `true`
- when no block or argument is given returns true when none of the collection members are false or nil:
 `p [1, true, 'hi', false].my_all?` should return `false`
- when a class is passed as an argument returns true if all of the collection is a member of such class
`p [1,1,1].my_all?(Integer)` should return `true`
-  when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['d','d'].my_all?(/d/)` should return `true`
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`p [3,3,3,3].my_all?(3)` should return `true`
- [ ] `my_any?`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p [2,32].my_any?('cat')` should return `false`
 - [ ] `my_none?`:
- returns true if the block never returns true for all elements
block = proc { |num| num < 2 }
`p [2,23,21,12].my_none?(&block)` should return `true` 
- When no block or argument is given returns true only if none of the collection members is true
`p  [false,false].my_none?` should return `true`
- When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`p ['hello'].my_none?(5)` - should return `true`
- when a class is passed as an argument returns true if none of the collection is a member of such class
`p [3,3,3,3].my_none?(String)`  should return `true`
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
`p ['hello','milk','milo'].my_none?(/z/)` should return `true`
- [ ] `my_map` 
- returns an Enumerator if no block is given:
  `p [1, 2, 3].my_map` returns `#<Enumerator>`
- [ ] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(:*)` should return `151200`
- Combines all elements of enum by applying a binary operation, specified by a block:
`search = proc { |memo, word| memo.length > word.length ? memo : word }`
`p ['hello','strong','am'].my_inject(&search)` should return `strong`


### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Require changes🛑
# Hi Appiah
Good job on implement almost all the changes requested in the previous review.
### Some of the issue raised in relation to ```my_inject```, ```my_select```, still persist
* [ ] ```p [1, 2, 3].my_select``` should return same results as ```p [1, 2, 3].select ```
* [ ] ```p (5..10).my_inject(:*)``` should return same results as ```p (5..10).inject(:*)```
* [ ] ```search = proc { |memo, word| memo.length > word.length ? memo : word }```
* [ ] ```['hello','strong','am'].my_inject(&search)``` should return same result as ```['hello','strong','am'].inject(&search)``` 
## Kindly address them and submit the project for review
### Happy coding💻
Addo

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @appiersign       
#### Great Job so far, but there are a few more things required. 
### General
- [ ] `my_all?`
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
`p ['d','d'].my_all?(/c/)` should return `false`
- [ ] `my_none?`
- When no block or argument is given returns true only if none of the collection members is true
p [false,true].my_none? should return false 
- [ ] `my_any?`
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`


### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @appiersign       
#### Great Job you are almost there,just few more things required. 
### General
- [ ] `my_none?`
- When no block or argument is given returns true only if none of the collection members is true
`p [false,false].my_none?` should return` true`
- [ ] `my_any?`
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, true, false].my_any?` should return `true`


### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+      my_each { |x| return false if x.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if x.match(arg) }
+    elsif arg && !arg.is_a?(Module) && !arg.is_a?(Regexp)
+      my_each { |x| return false if x == arg }
+    else
+      my_each { |x| none = true unless x }
+    end
+    none
+  end
```
Comment (maelfosso) [inline main.rb:88]:
```ruby
2.6.5 :009 > array
 => [8, 3, 4, 3, 6, 3, 5, 5, 4, 2, 4, 2, 1, 1, 5, 3, 8, 5, 4, 4, 6, 1, 2, 8, 8, 3, 4, 6, 3, 0, 3, 7, 1, 4, 2, 5, 8, 2, 6, 7, 8, 7, 5, 7, 6, 3, 2, 4, 0, 0, 5, 4, 7, 6, 8, 1, 2, 7, 7, 4, 6, 3, 3, 7, 8, 8, 0, 4, 8, 7, 2, 5, 0, 1, 1, 3, 6, 4, 4, 5, 0, 2, 0, 6, 6, 2, 1, 6, 7, 8, 0, 2, 7, 5, 2, 3, 3, 8, 5, 8] 
2.6.5 :010 > 
2.6.5 :011 > 
2.6.5 :012 > true_block = proc { |num| num > 9 }
 => #<Proc:0x000055563335c990@(irb):12> 
2.6.5 :013 > array.my_none?(&true_block)
 => false 
2.6.5 :014 > array.none?(&true_block)
 => true 
2.6.5 :015 > 
```

Comment (maelfosso) [reply main.rb:88]:
```ruby
2.6.5 :030 > true_array = [nil, false, true, []]
 => [nil, false, true, []] 
2.6.5 :031 > true_array.my_none?
 => true 
2.6.5 :032 > true_array.none?
 => false 
2.6.5 :033 > 
```

Comment (maelfosso) [reply main.rb:88]:
```ruby
2.6.5 :038 > array
 => [8, 3, 4, 3, 6, 3, 5, 5, 4, 2, 4, 2, 1, 1, 5, 3, 8, 5, 4, 4, 6, 1, 2, 8, 8, 3, 4, 6, 3, 0, 3, 7, 1, 4, 2, 5, 8, 2, 6, 7, 8, 7, 5, 7, 6, 3, 2, 4, 0, 0, 5, 4, 7, 6, 8, 1, 2, 7, 7, 4, 6, 3, 3, 7, 8, 8, 0, 4, 8, 7, 2, 5, 0, 1, 1, 3, 6, 4, 4, 5, 0, 2, 0, 6, 6, 2, 1, 6, 7, 8, 0, 2, 7, 5, 2, 3, 3, 8, 5, 8] 
2.6.5 :039 > array.my_none?(String)
 => false 
2.6.5 :040 > array.none?(String)
 => true 
2.6.5 :041 > 
```

Comment (maelfosso) [reply main.rb:88]:
```ruby
2.6.5 :068 > words = %w[dog door rod blade]
 => ["dog", "door", "rod", "blade"] 
2.6.5 :069 > words.my_none?(/z/)
 => false 
2.6.5 :070 > words.none?(/z/)
 => true 
2.6.5 :071 > 
```

Comment (maelfosso) [reply main.rb:88]:
```ruby
2.6.5 :073 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :074 > words.my_none?(5)
 => false 
2.6.5 :075 > words.none?(5)
 => true 
2.6.5 :076 > 
```

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      sym = symbol if symbol.is_a?(Symbol) || symbol.is_a?(String)
+    end
+
+    if sym
+      arr.my_each { |x| acc = acc.__send__(sym, x) }
+    elsif block_given?
+      arr.my_each { |x| acc = yield(acc, x) }
+    end
+    acc
+  end
```
Comment (maelfosso) [inline main.rb]:
```ruby
2.6.5 :086 > operation = proc { |sum, n| sum + n }
 => #<Proc:0x000055563344ca58@(irb):86> 
2.6.5 :087 > array
 => [8, 3, 4, 3, 6, 3, 5, 5, 4, 2, 4, 2, 1, 1, 5, 3, 8, 5, 4, 4, 6, 1, 2, 8, 8, 3, 4, 6, 3, 0, 3, 7, 1, 4, 2, 5, 8, 2, 6, 7, 8, 7, 5, 7, 6, 3, 2, 4, 0, 0, 5, 4, 7, 6, 8, 1, 2, 7, 7, 4, 6, 3, 3, 7, 8, 8, 0, 4, 8, 7, 2, 5, 0, 1, 1, 3, 6, 4, 4, 5, 0, 2, 0, 6, 6, 2, 1, 6, 7, 8, 0, 2, 7, 5, 2, 3, 3, 8, 5, 8] 
2.6.5 :088 > array.my_inject(&operation)
 => 425 
2.6.5 :089 > array.inject(&operation)
 => 417 
2.6.5 :090 > 
```

Comment (maelfosso) [reply main.rb]:
```ruby
2.6.5 :092 > array
 => [3, 4, 3, 6, 3, 5, 5, 4, 2, 4, 2, 1, 1, 5, 3, 8, 5, 4, 4, 6, 1, 2, 8, 8, 3, 4, 6, 3, 0, 3, 7, 1, 4, 2, 5, 8, 2, 6, 7, 8, 7, 5, 7, 6, 3, 2, 4, 0, 0, 5, 4, 7, 6, 8, 1, 2, 7, 7, 4, 6, 3, 3, 7, 8, 8, 0, 4, 8, 7, 2, 5, 0, 1, 1, 3, 6, 4, 4, 5, 0, 2, 0, 6, 6, 2, 1, 6, 7, 8, 0, 2, 7, 5, 2, 3, 3, 8, 5, 8] 
2.6.5 :093 > array.my_inject(:+)
 => 417 
2.6.5 :094 > array.inject(:+)
 => 414 
2.6.5 :095 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle:

Hello,
Thank you for the improvements. 
However, there are still some changes to do. Please, read the comments below.

Thanks,
Mael FOSSO

File: .rubocop.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+Style/Copyright:
+  Enabled: false
+Style/DocumentationMethod:
+  Enabled: false
+Style/InlineComment:
+  Enabled: false
+Style/MissingElse:
+  Enabled: false
+Style/MethodCalledOnDoEndBlock:
+  Enabled: false
```
Comment (shshamim63) [inline .rubocop.yml:52]: Why do you need to make changes to the rubcop.yml file? You should not alter it. If you need to disable any rules, you have to do so by inline comments that you have done already. Please read the instruction in [here](https://github.com/microverseinc/linters-config/tree/master/ruby).

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      arr_temp.my_each { |x| acc = yield(acc, x) }
+    else
+      arr_temp.my_each { |x| acc = acc.__send__(sym, x) }
+    end
+    acc
+  end
+
+  def multiply_els(array)
+    array.my_inject(1) { |p, x| p * x }
+  end
```
Comment (shshamim63) [inline main.rb]: *multiply_els* is not an enumerable method. So it should be placed outside of your module. The main purpose of this method is to check can you access your methods or not.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :traffic_light: 
___
Hi @appiersign :wave.
Congratulation :clap: , all of your methods are working properly. But have found some issues that are left behind and they are crucial. I am really sorry, they were not mentioned by the previous reviewers :disappointed: . But these are crucial.
___
### :spades: Mandatory Changes :spades: 
___
1. You should not change the linter config files. Check my comment below for more explanation.
2. Place the multiply_els in the right place. Check the comment below.
___
##### Please fix the pointed issued and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

File: main.rb (hunk 4 above)
Comment (appiersign) [reply main.rb]: The following cops were added to RuboCop, but are not configured. Please set Enabled to either `true` or `false` in your `.rubocop.yml` file:
 - Lint/RaiseException (0.81)
 - Lint/StructNewOverride (0.81)
 - Style/HashEachMethods (0.80)
 - Style/HashTransformKeys (0.80)
 - Style/HashTransformValues (0.80)

that's the linter error I get after replacing my rubocop file with the default one
if you're requesting me to to update it and rubocop if complaining when I try using inline
what do you suggest I do?

File: .rubocop.yml (hunk 3 above)
Comment (appiersign) [reply .rubocop.yml:52]: The following cops were added to RuboCop, but are not configured. Please set Enabled to either true or false in your .rubocop.yml file:

Lint/RaiseException (0.81)
Lint/StructNewOverride (0.81)
Style/HashEachMethods (0.80)
Style/HashTransformKeys (0.80)
Style/HashTransformValues (0.80)
that's the linter error I get after replacing my rubocop file with the default one
if you're requesting me to to update it and rubocop if complaining when I try using inline
what do you suggest I do?

**Review by Forison — CHANGES_REQUESTED** [review]
### Status::Require changes🛑
# 🙋‍♂️
The issues raised in the previous review still persist, Kindly address it.
```rubocop.yml``` should be the same as Microverse's ```rubocop.yml```
## Kindly make the changes and submit the project for review.
### Happy coding
Addo

**Comment by appiersign** [comment]
Can you be specific

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello @appiersign  ,  You have done a great job with this project. :clap:

**Pull request merged**